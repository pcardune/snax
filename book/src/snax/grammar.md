<script src="/main.js"></script>

# The Snax Grammar

<script>
  nfaExplore.render.SnaxGrammar({});
</script>
