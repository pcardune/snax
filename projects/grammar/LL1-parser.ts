import { err, ok, Result } from 'neverthrow';
import { LexToken } from '../lexer-gen/lexer-gen';
import {
  ActionFunction,
  BacktrackFreeGrammar,
  Eof,
  EOF,
  EPSILON,
  Epsilon,
  Production,
} from './grammar';

type LL1Table<S, A> = Map<S, Map<S | Eof | Epsilon, null | Production<S, A>>>;

class LL1TableImpl<S, A> {
  private table: LL1Table<S, A> = new Map();

  set(
    nonTerminal: S,
    terminal: S | Eof | Epsilon,
    cell: null | Production<S, A>
  ) {
    let columns = this.table.get(nonTerminal);
    if (!columns) {
      columns = new Map();
    }
    columns.set(terminal, cell);
    this.table.set(nonTerminal, columns);
  }

  get(nonTerminal: S, terminal: S | Eof | Epsilon) {
    return this.table.get(nonTerminal)?.get(terminal) || null;
  }

  *entries(): Generator<[S, S | Eof | Epsilon, null | Production<S, A>]> {
    for (const [row, cols] of this.table.entries()) {
      for (const [col, cell] of cols.entries()) {
        yield [row, col, cell] as [
          S,
          S | Eof | Epsilon,
          null | Production<S, A>
        ];
      }
    }
  }
}

/**
 * Algorithm for construction of LL1 table from backtrack free grammar.
 * See page 113 of Engineering a Compiler 2nd Edition
 */
export function buildLL1Table<S, A>(
  grammar: BacktrackFreeGrammar<S, A>
): LL1TableImpl<S, A> {
  const table: LL1TableImpl<S, A> = new LL1TableImpl();

  const terminals = new Set(grammar.getTerminals());
  terminals.delete(EPSILON as any);
  for (const A of grammar.getNonTerminals()) {
    for (const w of [...terminals, EOF] as (S | Eof)[]) {
      table.set(A, w, null);
    }
    for (const p of grammar.productionsFrom(A)) {
      const firstPlus = grammar.getFirstPlus(p);
      for (const w of firstPlus) {
        if (terminals.has(w as S)) {
          table.set(A, w, p);
        }
      }
      if (firstPlus.has(EOF)) {
        table.set(A, EOF, p);
      }
    }
  }
  return table;
}

export class LL1Parser<S, A> {
  readonly grammar: BacktrackFreeGrammar<S, A>;
  private table: Readonly<LL1TableImpl<S, A>>;
  private start: S;
  constructor(grammar: BacktrackFreeGrammar<S, A>, start: S) {
    this.grammar = grammar;
    this.table = buildLL1Table(this.grammar);
    this.start = start;
  }

  parseOrThrow(tokens: string[]) {
    const result = this.parse(tokens);
    if (result.isOk()) {
      return result.value;
    }
    throw result.error;
  }
  parse(tokens: string[]) {
    return this.parseTokens(
      tokens.map((t) => new LexToken(t as unknown as S, { from: 0, to: 0 }, t))
    );
  }
  parseTokensOrThrow(tokens: Iterable<LexToken<S>>): A | null {
    const r = this.parseTokens(tokens);
    if (r.isOk()) {
      return r.value;
    }
    throw r.error;
  }
  /**
   * Implements table driven LL(1) skeleton parser as described on
   * page 112 of Engineering a Compiler 2nd Edition.
   * @param tokens
   * @returns
   */
  parseTokens(tokens: Iterable<LexToken<S>>): Result<A | null, any> {
    const callAction = (
      action: ActionFunction<A>,
      tokens: (LexToken<S | Eof> | A)[]
    ) => {
      const nodes = tokens.map((n) => (n instanceof LexToken ? undefined : n));
      const onlyTokens = tokens.map((t) =>
        t instanceof LexToken ? t : undefined
      );
      const value = action(nodes as A[], onlyTokens as LexToken<S>[]);
      return value;
    };

    const tokensIter = tokens[Symbol.iterator]();
    const nextWord = () => {
      const next = tokensIter.next();
      if (!next.done) {
        return next.value;
      }
      return new LexToken(EOF, { from: 0, to: 0 }, '');
    };
    const terminals = this.grammar.getTerminals();
    let word = nextWord();
    type State = {
      symbol: S | Eof;
      collect?: boolean;
      production?: Production<S, A>;
    };
    let stack: State[] = [{ symbol: EOF, collect: false }];
    stack.push({ symbol: this.start, collect: false });
    let focus = stack[stack.length - 1];
    let root: A | null = null;
    let collected: (LexToken<S | Eof> | A)[] = [];
    while (true) {
      if (focus.collect) {
        stack.pop();
        if (!focus.production) {
          throw new Error(
            `Unexpected lack of a production to collect for ${String(
              focus.symbol
            )}`
          );
        }
        const { action, symbols } = focus.production;
        const tokens = collected.slice(collected.length - symbols.length);
        collected = collected.slice(0, collected.length - symbols.length);
        collected.push(callAction(action, tokens));
      } else if (focus.symbol === EOF && word.token === EOF) {
        // success
        return ok(collected[0] as A);
      } else if (focus.symbol === EOF || terminals.has(focus.symbol)) {
        if (focus.symbol === word.token) {
          collected.push(word);
          stack.pop();
          word = nextWord();
        } else {
          return err(new Error(`Couldn't find symbol ${String(focus.symbol)}`));
        }
      } else {
        // focus is a non-terminal
        let production = this.table.get(focus.symbol, word.token);
        if (production) {
          const A = production.rule;
          const B = production.symbols;
          focus.collect = true;
          focus.production = production;
          for (let i = B.length - 1; i >= 0; i--) {
            if (B[i] !== (EPSILON as any)) {
              stack.push({ symbol: B[i], collect: false });
            }
          }
        } else {
          return err(new Error(`Failed to expand ${focus.symbol}`));
        }
      }
      focus = stack[stack.length - 1];
    }
  }
}
