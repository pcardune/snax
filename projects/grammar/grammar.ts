import { LexToken } from '../lexer-gen/lexer-gen';
import { OrderedMap } from '../utils/data-structures/OrderedMap';
import { ParseNode } from './top-down-parser';

export const EPSILON = Symbol('ϵ');
export const EOF = Symbol('EOF');
export type Epsilon = typeof EPSILON;
export type Eof = typeof EOF;

export interface ConstGrammar<Symbol, ActionValue = any> {
  productionsIter(): Generator<Production<Symbol, ActionValue>>;
  productionsFrom(rule: Symbol): Readonly<Production<Symbol, ActionValue>[]>;
  getNonTerminals(): Readonly<Symbol[]>;
  getTerminals(): Readonly<Set<Symbol>>;
  toString(): string;
}

export type ActionFunction<ActionValue> = (
  childValues: ActionValue[],
  tokens: LexToken<unknown>[]
) => ActionValue;

// TODO: make specific generic type for terminal
// and non terminal symbols
export class Grammar<Symbol, ActionValue = any>
  implements ConstGrammar<Symbol, ActionValue>
{
  private productions: OrderedMap<Symbol, Production<Symbol, ActionValue>[]> =
    new OrderedMap();

  createProduction(
    rule: Symbol,
    symbols: (Symbol | SemanticAction<ActionValue>)[],
    action?: ActionFunction<ActionValue>
  ) {
    if (action) {
      symbols.push(new SemanticAction(action));
    }
    const production: Production<Symbol, ActionValue> = new Production(
      rule,
      symbols
    );
    this.addProduction(production);
    return production;
  }
  addProduction(production: Production<Symbol, ActionValue>) {
    const key = production.rule;
    if (!this.productions.get(key)) {
      this.productions.set(key, []);
    }
    this.productions.get(key)?.push(production);
  }

  removeProduction(production: Production<Symbol>) {
    const key = production.rule;
    const prods = this.productions.get(key);
    if (prods) {
      this.productions.set(
        key,
        prods.filter((p) => !p.equals(production))
      );
      if (this.productions.get(key)?.length == 0) {
        this.productions.delete(key);
      }
    }
  }

  addProductions(rule: Symbol, derivations: Symbol[][]) {
    for (const derivation of derivations) {
      this.addProduction(new Production(rule, derivation));
    }
  }
  *productionsIter() {
    for (const key of this.productions.keys()) {
      for (const production of this.productions.get(key) || []) {
        yield production;
      }
    }
  }
  productionsFrom(rule: Symbol): Readonly<Production<Symbol, ActionValue>[]> {
    return this.productions.get(rule as Symbol) || [];
  }
  getNonTerminals(): Readonly<Symbol[]> {
    return [...this.productions.keys()];
  }
  getTerminals(): Readonly<Set<Symbol>> {
    let terminals: Set<Symbol> = new Set();
    for (const value of this.productions.values()) {
      for (const p of value) {
        for (const s of p.symbols) {
          if (!this.productions.get(s)) {
            terminals.add(s);
          }
        }
      }
    }
    return terminals;
  }
  toString() {
    let out = '\n';
    let terminals = this.getTerminals();
    for (const productions of this.productions.values()) {
      out += `${productions[0].rule} →\n`;
      for (const production of productions) {
        out += '  | ';
        out += production.symbols
          .map((s) => {
            if ((s as any) === EPSILON) {
              return 'ϵ';
            }
            return terminals.has(s) ? `'${s}'` : s;
          })
          .join(' ');
        out += '\n';
      }
    }
    return out;
  }
}

class SemanticAction<ActionValue> {
  func: ActionFunction<ActionValue>;
  constructor(func: ActionFunction<ActionValue>) {
    this.func = func;
  }
}

export class Production<Symbol, ActionValue = void> {
  /**
   * The left hand symbol. So the production:
   *   Expr -> Term Op Term
   * the `rule` would be `Expr`
   */
  readonly rule: Symbol;

  /**
   * The right hand side of a production. So for the production:
   *   Expr -> Term Op Term
   * the `symbols` would be `[Term, Op, Term]`
   */
  readonly body: Readonly<(Symbol | SemanticAction<ActionValue>)[]>;

  constructor(
    rule: Symbol,
    symbols: (Symbol | SemanticAction<ActionValue>)[],
    action?: ActionFunction<ActionValue>
  ) {
    if (action) {
      symbols.push(new SemanticAction(action));
    }
    this.rule = rule;
    this.body = symbols;
  }

  get symbols(): Symbol[] {
    return this.body.filter((s) => !(s instanceof SemanticAction)) as Symbol[];
  }

  /**
   * @deprecated
   */
  get action(): ActionFunction<ActionValue> {
    const action = this.body.find((s) => s instanceof SemanticAction) as
      | SemanticAction<ActionValue>
      | undefined;
    if (action) {
      return action.func;
    }
    return () => {
      throw new Error('No action for production ' + this.toString());
    };
  }

  isLeftRecursive(): boolean {
    return (this.symbols[0] as any) === this.rule;
  }

  equals(other: Production<any>): boolean {
    if (this.rule != other.rule) {
      return false;
    }
    if (this.symbols.length != other.symbols.length) {
      return false;
    }
    for (let i = 0; i < this.symbols.length; i++) {
      if (!this.symbols[i] === other.symbols[i]) {
        return false;
      }
    }
    return true;
  }

  toString(): string {
    return `${this.rule} -> ${this.symbols.map((s) => String(s)).join(' ')}`;
  }
}
export type ParseNodeGrammar = Grammar<
  string | typeof EPSILON,
  ParseNode<string | typeof EPSILON | null, LexToken<any>>
>;
export type GrammarSpec = { [index: string]: string[][] };
export function buildGrammar(productions: GrammarSpec) {
  type GrammarSymbol = string | typeof EPSILON;

  const grammar: ParseNodeGrammar = new Grammar();
  const nonTerminals: Set<Symbol> = new Set();
  Object.keys(productions).forEach((key) => {
    nonTerminals.add(key as unknown as Symbol);
  });
  Object.entries(productions).forEach(([key, value]) => {
    const symbol = key;
    for (const specSymbols of value) {
      if (specSymbols.length === 0) {
        // this is an empty rule, put epsilon in it
        grammar.createProduction(symbol, [
          EPSILON,
          new SemanticAction(
            (children, tokens) => new ParseNode(symbol, children)
          ),
        ]);
      } else {
        grammar.createProduction(symbol, [
          ...specSymbols,
          new SemanticAction((children, tokens) => {
            let spliced: ParseNode<GrammarSymbol | null, LexToken<any>>[] = [];
            for (let [i, child] of children.entries()) {
              if (child) {
                spliced.push(child);
              } else {
                spliced.push(ParseNode.forToken(tokens[i]));
              }
            }
            return new ParseNode(symbol, spliced);
          }),
        ]);
      }
    }
  });
  return grammar;
}

function setsAreEqual(s1: ReadonlySet<any>, s2: ReadonlySet<any>) {
  if (s1.size !== s2.size) {
    return false;
  }
  for (const s of s1) {
    if (!s2.has(s)) {
      return false;
    }
  }
  return true;
}

type FirstSet<S> = ReadonlySet<S | Eof | Epsilon>;
type FirstMap<S> = Map<S | Eof | Epsilon, FirstSet<S>>;
/**
 * Algorithm to calculate the set of terminal symbols
 * that can appear as the first word in some string of symbols.
 * @returns a map from symbols in the given grammar to their "first" sets
 *
 * See Page 104 of Engineering a Compiler 2nd Edition
 */
export function calcFirst<S>(grammar: ConstGrammar<S>): FirstMap<S> {
  const firstMap: FirstMap<S> = new Map();

  let done = false;
  const setFirst = (k: S, newFirst: FirstSet<S>) => {
    const existingFirst = getFirst(k);
    firstMap.set(k, newFirst);
    if (!setsAreEqual(existingFirst, newFirst)) {
      done = false;
    }
  };
  const getFirst = (k: S) => {
    if (!firstMap.has(k)) {
      firstMap.set(k, new Set());
    }
    return firstMap.get(k) as FirstSet<S>;
  };

  for (const terminal of grammar.getTerminals()) {
    firstMap.set(terminal, new Set([terminal]));
    firstMap.set(EPSILON, new Set([EPSILON]));
    firstMap.set(EOF, new Set([EOF]));
  }

  while (!done) {
    done = true;
    for (const production of grammar.productionsIter()) {
      const B = production.symbols;
      const b1 = B[0];
      let rhs: Set<S | Eof | Epsilon> = new Set([...getFirst(b1)]);
      rhs.delete(EPSILON);
      let i = 0;
      let k = production.symbols.length - 1;
      while (getFirst(B[i]).has(EPSILON) && i <= k - 1) {
        for (const s of getFirst(B[i + 1])) {
          rhs.add(s);
        }
        rhs.delete(EPSILON);
        i++;
      }
      if (i == k && getFirst(B[k]).has(EPSILON)) {
        rhs.add(EPSILON);
      }
      const A = production.rule;
      const fA = new Set([...getFirst(A), ...rhs]);
      setFirst(A, fA);
    }
  }
  return firstMap;
}

/**
 * Calculate follow sets as described on page 106 of
 * Engineering a Compiler 2nd Edition
 *
 * @param grammar a grammar
 * @param firstMap the first sets calculated with {@link calcFirst}
 * @returns a mapping from non terminal symbols in the given grammar
 * to their follow sets
 */
export function calcFollow<S>(
  grammar: ConstGrammar<S>,
  firstMap: FirstMap<S>,
  start: S
): FirstMap<S> {
  const getFirst = (k: S) => firstMap.get(k) as FirstSet<S>;

  const followMap: FirstMap<S> = new Map();
  const nonTerminals = new Set(grammar.getNonTerminals());
  for (const nonTerminal of nonTerminals) {
    followMap.set(nonTerminal, new Set());
  }
  followMap.set(start, new Set([EOF]));
  let done = false;
  const setFollow = (k: S, newFollow: FirstSet<S>) => {
    const existingFirst = getFollow(k);
    followMap.set(k, newFollow);
    if (!setsAreEqual(existingFirst, newFollow)) {
      done = false;
    }
  };
  const getFollow = (k: S) => {
    return followMap.get(k) as FirstSet<S>;
  };

  while (!done) {
    done = true;
    for (const production of grammar.productionsIter()) {
      let B = production.symbols;
      let k = B.length - 1;
      let trailer: Set<S | Eof | Epsilon> = new Set(getFollow(production.rule));
      for (let i = k; i >= 0; i--) {
        if (nonTerminals.has(B[i])) {
          setFollow(B[i], new Set([...getFollow(B[i]), ...trailer]));
          const firstBi = getFirst(B[i]);
          if (firstBi.has(EPSILON as any)) {
            trailer = new Set([...trailer, ...firstBi]);
            trailer.delete(EPSILON as any);
          } else {
            trailer = new Set(firstBi);
          }
        } else {
          trailer = new Set(getFirst(B[i]));
        }
      }
    }
  }
  return followMap;
}

function intersection<S>(a: ReadonlySet<S>, b: ReadonlySet<S>): Set<S> {
  const result: Set<S> = new Set();
  for (const i of a) {
    if (b.has(i)) {
      result.add(i);
    }
  }
  return result;
}

export function isBacktrackFree<S>(grammar: ConstGrammar<S>, start: S) {
  try {
    new BacktrackFreeGrammar(grammar, start);
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * Tests whether a given array has a given prefix
 *
 * @param string array to test
 * @param prefix prefix of the array
 * @returns whether or not prefix is a prefix of string
 */
export function startsWith<S>(
  string: readonly S[],
  prefix: readonly S[]
): boolean {
  if (string.length < prefix.length) {
    return false;
  }
  for (let i of prefix.keys()) {
    if (string[i] != prefix[i]) {
      return false;
    }
  }
  return true;
}

/**
 * Given a list of strings, find the longest prefix that matches
 * more than one of those strings.
 *
 * @param strings
 * @returns
 */
export function findLongestPrefix<S>(strings: readonly (readonly S[])[]): S[] {
  if (strings.length <= 1) {
    return [];
  }
  let longestPrefix: S[] = [];
  for (const [i, si] of strings.entries()) {
    if (si.length === 0) {
      continue;
    }
    for (let j = si.length; j > 0; j--) {
      let prefix = si.slice(0, j);
      let matches = strings.filter((m) => startsWith(m, prefix));
      if (matches.length > 1 && prefix.length > longestPrefix.length) {
        longestPrefix = prefix;
      }
    }
  }
  return longestPrefix;
}

/**
 * Left factoring algorithm for grammars to make them backtrack free
 * Described on Page 108 of Engineering a Compiler 2nd Edition.
 */
export function leftFactor<S, A>(
  grammar: Grammar<S, A>
): Grammar<S | string, A> {
  const newGrammar: Grammar<S | string, A> = new Grammar();

  const nonTerminals = grammar.getNonTerminals();

  let newSymbolCount = 0;
  function createSymbol(): string {
    return `__left_factor_${newSymbolCount++}`;
  }

  for (const A of nonTerminals) {
    const productions = grammar.productionsFrom(A);
    if (productions.length == 1) {
      // no need to left factor productions with one right-hand-side
      newGrammar.addProduction(productions[0]);
      continue;
    }

    const prefix = findLongestPrefix(productions.map((p) => p.symbols));
    if (prefix.length == 0) {
      for (const p of productions) {
        newGrammar.addProduction(p);
      }
    } else {
      // now we need to create a new production.
      const newRule = createSymbol();
      newGrammar.addProduction(new Production(A, [...prefix, newRule]));
      for (const p of productions) {
        if (startsWith(p.symbols, prefix)) {
          const suffix: (S | string)[] = p.symbols.slice(prefix.length);
          newGrammar.addProduction(new Production(newRule, suffix));
        } else {
          newGrammar.addProduction(p);
        }
      }
    }
  }
  return newGrammar;
}

/**
 * A wrapper around Grammar that provides additional information
 * unique to backtrack free grammars.
 */
export class BacktrackFreeGrammar<S, A> implements ConstGrammar<S, A> {
  private readonly grammar: ConstGrammar<S, A>;
  private readonly firstMap: FirstMap<S>;
  private readonly followMap: FirstMap<S>;
  constructor(grammar: ConstGrammar<S, A>, start: S) {
    this.grammar = grammar;
    this.firstMap = calcFirst(this.grammar);
    this.followMap = calcFollow(this.grammar, this.firstMap, start);
    if (!this.isValid()) {
      throw new Error('Grammar is not backtrack free...');
    }
  }

  productionsIter(): Generator<Production<S, A>, any, unknown> {
    return this.grammar.productionsIter();
  }

  productionsFrom(rule: S): readonly Production<S, A>[] {
    return this.grammar.productionsFrom(rule);
  }

  getNonTerminals(): readonly S[] {
    return this.grammar.getNonTerminals();
  }

  getTerminals(): Readonly<Set<S>> {
    return this.grammar.getTerminals();
  }

  toString(): string {
    return this.grammar.toString();
  }

  getFirst(B: S | Readonly<S[]>): FirstSet<S> {
    if (!(B instanceof Array)) {
      return this.firstMap.get(B) as FirstSet<S>;
    }
    let first: Set<S | Epsilon | Eof> = new Set();
    for (let i = 0; i < B.length; i++) {
      const Bi = B[i];
      const firstBi = this.firstMap.get(Bi) as FirstSet<S>;
      first = new Set([...first, ...firstBi]);
      if (!firstBi.has(EPSILON)) {
        break;
      }
    }
    return first;
  }

  getFirstPlus(production: Production<S, A>): FirstSet<S> {
    const A = production.rule;
    const B = production.symbols;
    const firstB = this.getFirst(B);
    if (firstB.has(EPSILON)) {
      const followA = this.followMap.get(A) as FirstSet<S>;
      return new Set([...firstB, ...followA]);
    } else {
      return firstB;
    }
  }

  /**
   * Detect whether or not a grammar is backtrack free, according
   * to the definition on Page 107 of Engineering a Compiler 2nd Edition
   *
   * @param grammar a Grammar
   */
  isValid() {
    for (const A of this.getNonTerminals()) {
      const B = this.productionsFrom(A);
      for (let i = 0; i < B.length; i++) {
        const firstPlusBi = this.getFirstPlus(B[i]);
        for (let j = 0; j < B.length; j++) {
          if (i !== j) {
            const firstPlusBj = this.getFirstPlus(B[j]);
            if (intersection(firstPlusBi, firstPlusBj).size > 0) {
              // not backtrack free
              return false;
            }
          }
        }
      }
    }
    return true;
  }
}
