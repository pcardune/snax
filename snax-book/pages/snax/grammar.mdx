import {SnaxGrammar} from '../../components/SnaxGrammar.js';

# The Snax Grammar

<SnaxGrammar/>
