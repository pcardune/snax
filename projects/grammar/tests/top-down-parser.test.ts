import { buildGrammar, Grammar } from '../grammar';
import {
  buildParser,
  ParseErrorType,
  Parser,
  removeDirectLeftRecursion,
  removeLeftRecursion,
} from '../top-down-parser';

describe('removeDirectLeftRecursion', () => {
  const grammar: Grammar<string> = new Grammar();
  grammar.addProductions('Expr', [
    ['Expr', '+', 'Term'],
    ['Expr', '-', 'Term'],
    ['Term'],
  ]);
  grammar.addProductions('Term', [
    ['Term', '*', 'Factor'],
    ['Term', '/', 'Factor'],
    ['Factor'],
  ]);
  grammar.addProductions('Factor', [['(', 'Expr', ')'], ['num'], ['name']]);
  test('it works', () => {
    removeDirectLeftRecursion(grammar);
    expect(grammar.toString()).toMatchInlineSnapshot(`
"
Factor →
  | '(' Expr ')'
  | 'num'
  | 'name'
ExprP →
  | '+' Term ExprP
  | '-' Term ExprP
  | ϵ
Expr →
  | Term ExprP
TermP →
  | '*' Factor TermP
  | '/' Factor TermP
  | ϵ
Term →
  | Factor TermP
"
`);
  });
});

describe('removeLeftRecursion', () => {
  const grammar: Grammar<string> = new Grammar();
  grammar.addProductions('A', [['B']]);
  grammar.addProductions('B', [['C']]);
  grammar.addProductions('C', [['A', 'd']]);
  test('it works', () => {
    removeLeftRecursion(grammar);
    expect(grammar.toString()).toMatchInlineSnapshot(`
"
A →
  | B
B →
  | C
CP →
  | 'd' CP
  | ϵ
C →
  | CP
"
`);
  });
});

describe('parse', () => {
  describe('simple grammar', () => {
    // Root -> num
    const grammar = buildGrammar({ Root: [['num']] });
    const parser = new Parser(grammar, 'Root');
    test('parse', () => {
      const tree = parser.parseOrThrow(['num']);
      expect(tree.pretty()).toMatchInlineSnapshot(`
        "
        <Root>
        |  <num>num</num>
        </Root>
        "
      `);
      expect(() => parser.parseOrThrow(['bad'])).toThrow();
      expect(() => parser.parseOrThrow([])).toThrow();
    });
  });

  describe('sequence grammar', () => {
    const parser = buildParser({
      Root: [['first', 'second', 'third']],
    });
    test('parse', () => {
      const tokens = ['first', 'second', 'third'];
      expect(parser.parseOrThrow(tokens).pretty()).toMatchInlineSnapshot(`
        "
        <Root>
        |  <first>first</first>
        |  <second>second</second>
        |  <third>third</third>
        </Root>
        "
      `);
      expect(() => parser.parseOrThrow(['second', 'third', 'first'])).toThrow();
      expect(() => parser.parseOrThrow(['first', 'second'])).toThrow();
      expect(() =>
        parser.parseOrThrow(['first', 'second', 'third', 'first'])
      ).toThrow();
    });
  });

  describe('nested grammar', () => {
    const parser = buildParser({
      Root: [
        ['Child1', 'after-child-1'],
        ['Child2', 'after-child-2'],
      ],
      Child1: [['child-1']],
      Child2: [['child-2']],
    });
    test('parse', () => {
      expect(parser.parseOrThrow(['child-1', 'after-child-1']).pretty())
        .toMatchInlineSnapshot(`
        "
        <Root>
        |  <Child1>
        |  |  <child-1>child-1</child-1>
        |  </Child1>
        |  <after-child-1>after-child-1</after-child-1>
        </Root>
        "
      `);
      expect(parser.parseOrThrow(['child-2', 'after-child-2']).pretty())
        .toMatchInlineSnapshot(`
        "
        <Root>
        |  <Child2>
        |  |  <child-2>child-2</child-2>
        |  </Child2>
        |  <after-child-2>after-child-2</after-child-2>
        </Root>
        "
      `);
    });
  });

  describe('Putting back tokens while backtracking', () => {
    const parser = buildParser({
      Root: [
        ['before-A', 'A', 'after-A'],
        ['before-B', 'B', 'after-B'],
      ],
      A: [['a'], ['an']],
      B: [['be'], ['been']],
    });
    test('parses', () => {
      expect(parser.parseOrThrow(['before-B', 'been', 'after-B']).pretty())
        .toMatchInlineSnapshot(`
        "
        <Root>
        |  <before-B>before-B</before-B>
        |  <B>
        |  |  <been>been</been>
        |  </B>
        |  <after-B>after-B</after-B>
        </Root>
        "
      `);
    });
  });

  describe('complex grammar', () => {
    const parser = buildParser({
      Root: [['Expr']],
      Expr: [['Expr', '+', 'Term'], ['Expr', '-', 'Term'], ['Term']],
      Term: [['Term', '*', 'Factor'], ['Term', '/', 'Factor'], ['Factor']],
      Factor: [['(', 'Expr', ')'], ['num'], ['name']],
    });

    // for the expression 3, we would have the token "num";
    test('num', () => {
      const tokens = ['num'];
      const result = parser.parseOrThrow(tokens);
      expect(result.pretty()).toMatchInlineSnapshot(`
        "
        <Root>
        |  <Expr>
        |  |  <Term>
        |  |  |  <Factor>
        |  |  |  |  <num>num</num>
        |  |  |  </Factor>
        |  |  |  <Term>
        |  |  |  </Term>
        |  |  </Term>
        |  |  <Expr>
        |  |  </Expr>
        |  </Expr>
        </Root>
        "
      `);
    });
    test('num+num', () => {
      const tokens = ['num', '+', 'num'];
      const result = parser.parseOrThrow(tokens);
      expect(result.pretty()).toMatchSnapshot();
    });
    test('(num+num)', () => {
      const tokens = ['(', 'num', '+', 'num', ')'];
      const result = parser.parseOrThrow(tokens);
      expect(result.pretty()).toMatchSnapshot();
    });
    test('num+num+(num-num)', () => {
      const tokens = ['num', '+', 'num', '+', '(', 'num', '-', 'num', ')'];
      const result = parser.parseOrThrow(tokens);
      expect(result.pretty()).toMatchSnapshot();
    });
    describe('error handling', () => {
      test('+ should not parse', () => {
        const tokens = ['+'];
        const result = parser.parse(tokens);
        if (result.isOk()) {
          fail();
        }
        expect(result.error.type).toBe(ParseErrorType.TOKENS_REMAIN);
      });
      test('num+ should not parse', () => {
        const tokens = ['num', '+'];
        const result = parser.parse(tokens);
        if (result.isOk()) {
          fail();
        }
        expect(result.error.type).toBe(ParseErrorType.TOKENS_REMAIN);
      });
    });
  });
});
