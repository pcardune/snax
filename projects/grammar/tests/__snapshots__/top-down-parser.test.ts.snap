// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse complex grammar (num+num) 1`] = `
"
<Root>
|  <Expr>
|  |  <Term>
|  |  |  <Factor>
|  |  |  |  <(>(</(>
|  |  |  |  <Expr>
|  |  |  |  |  <Term>
|  |  |  |  |  |  <Factor>
|  |  |  |  |  |  |  <num>num</num>
|  |  |  |  |  |  </Factor>
|  |  |  |  |  |  <Term>
|  |  |  |  |  |  </Term>
|  |  |  |  |  </Term>
|  |  |  |  |  <Expr>
|  |  |  |  |  |  <+>+</+>
|  |  |  |  |  |  <Term>
|  |  |  |  |  |  |  <Factor>
|  |  |  |  |  |  |  |  <num>num</num>
|  |  |  |  |  |  |  </Factor>
|  |  |  |  |  |  |  <Term>
|  |  |  |  |  |  |  </Term>
|  |  |  |  |  |  </Term>
|  |  |  |  |  |  <Expr>
|  |  |  |  |  |  </Expr>
|  |  |  |  |  </Expr>
|  |  |  |  </Expr>
|  |  |  |  <)>)</)>
|  |  |  </Factor>
|  |  |  <Term>
|  |  |  </Term>
|  |  </Term>
|  |  <Expr>
|  |  </Expr>
|  </Expr>
</Root>
"
`;

exports[`parse complex grammar num+num 1`] = `
"
<Root>
|  <Expr>
|  |  <Term>
|  |  |  <Factor>
|  |  |  |  <num>num</num>
|  |  |  </Factor>
|  |  |  <Term>
|  |  |  </Term>
|  |  </Term>
|  |  <Expr>
|  |  |  <+>+</+>
|  |  |  <Term>
|  |  |  |  <Factor>
|  |  |  |  |  <num>num</num>
|  |  |  |  </Factor>
|  |  |  |  <Term>
|  |  |  |  </Term>
|  |  |  </Term>
|  |  |  <Expr>
|  |  |  </Expr>
|  |  </Expr>
|  </Expr>
</Root>
"
`;

exports[`parse complex grammar num+num+(num-num) 1`] = `
"
<Root>
|  <Expr>
|  |  <Term>
|  |  |  <Factor>
|  |  |  |  <num>num</num>
|  |  |  </Factor>
|  |  |  <Term>
|  |  |  </Term>
|  |  </Term>
|  |  <Expr>
|  |  |  <+>+</+>
|  |  |  <Term>
|  |  |  |  <Factor>
|  |  |  |  |  <num>num</num>
|  |  |  |  </Factor>
|  |  |  |  <Term>
|  |  |  |  </Term>
|  |  |  </Term>
|  |  |  <Expr>
|  |  |  |  <+>+</+>
|  |  |  |  <Term>
|  |  |  |  |  <Factor>
|  |  |  |  |  |  <(>(</(>
|  |  |  |  |  |  <Expr>
|  |  |  |  |  |  |  <Term>
|  |  |  |  |  |  |  |  <Factor>
|  |  |  |  |  |  |  |  |  <num>num</num>
|  |  |  |  |  |  |  |  </Factor>
|  |  |  |  |  |  |  |  <Term>
|  |  |  |  |  |  |  |  </Term>
|  |  |  |  |  |  |  </Term>
|  |  |  |  |  |  |  <Expr>
|  |  |  |  |  |  |  |  <->-</->
|  |  |  |  |  |  |  |  <Term>
|  |  |  |  |  |  |  |  |  <Factor>
|  |  |  |  |  |  |  |  |  |  <num>num</num>
|  |  |  |  |  |  |  |  |  </Factor>
|  |  |  |  |  |  |  |  |  <Term>
|  |  |  |  |  |  |  |  |  </Term>
|  |  |  |  |  |  |  |  </Term>
|  |  |  |  |  |  |  |  <Expr>
|  |  |  |  |  |  |  |  </Expr>
|  |  |  |  |  |  |  </Expr>
|  |  |  |  |  |  </Expr>
|  |  |  |  |  |  <)>)</)>
|  |  |  |  |  </Factor>
|  |  |  |  |  <Term>
|  |  |  |  |  </Term>
|  |  |  |  </Term>
|  |  |  |  <Expr>
|  |  |  |  </Expr>
|  |  |  </Expr>
|  |  </Expr>
|  </Expr>
</Root>
"
`;
